src/database/database-thread.c

```diff
@@ -135,16 +160,30 @@ void *DB_thread(void *val)
 			set_event(PARSE_NEIGHBOR_CACHE);
 		}
 
+		// Intermediate cancellation-point
+		if(killed)
+			break;
+
+		// Optimize database once per week
+		if(now - lastAnalyze >= 604800)
+		{
+			DBOPEN_OR_AGAIN();
+			analyze_database(db);
+			lastAnalyze = now;
+			DBCLOSE_OR_BREAK();
+		}
+
 		// Intermediate cancellation-point
 		if(killed)
 			break;
 
 		// Update MAC vendor strings once a month (the MAC vendor
```

Comment:
Is this necessary at all? We only download the vendor database on install/update which will trigger a FTL restart anyway  reading the database. 

P.S. How often is the vendor database updated at all? Maybe it's time to do it again?

Reply:
This is *not* done on an FTL restart, it is only done once per month.

> How often is the vendor database updated at all? Maybe it's time to do it again?

There is a `cron` job running on `ftl.pi-hole.net` doing this in frequent intervals, have a look at the last modified timestamp online:
![Screenshot from 2024-01-06 09-03-16](https://github.com/pi-hole/FTL/assets/16748619/b4916029-1fdf-4147-9eae-0929232daa15)

I will push another small change that will ensure the steps are run once per week/month (new behavior) and not after FTL has been running continuously for one week/month (current behavior).

Reply:
Pushed

Reply:
Where do you save the lastMACVendor between restarts? If it's not stored externally, you can't be sure the update is executed every month if FTL is restarted in-between.

Reply:
This is addressed by the last commit.

The update is started randomly within the first hour of the next month.

Example: If FTL is (re)started on the last day (31st), the update will still be done in *the next* month (one day after the restart in this example).

Edge-cases are considered: If FTL is offline in the first hour of say February, the first update will be performed on March 1st. When FTL is restarted within the first hour of March 1st (and before the update happens), the update is done in April.

Analyze is similar but within the first hour of each *week*.

Note: FTL's calculation is simplified and doesn't calculate months in a calendar sense but in a time sense, i.e., the first day of a month may actually be the 6th or so. This is because leap years continue to mix things up since the beginning of time (January 1st, 1970). In the end, it doesn't matter...